### PR TITLE
Prometheus functionality

### PR DESCRIPTION
**Pull request by jtlisi**
This is a cleaned up version of tsdb-gw with the prometheus code behind the macaron handler so basic auth can be used. I tested it locally but not extensively.

### CODE REVIEW

**Comment by jtlisi** [comment]
Let me rebase this before review

File: api/api.go [hunk 1]
```diff
@@ -95,6 +95,7 @@ func (a *Api) InitRoutes(m *macaron.Macaron) {
 	m.Get("/metrics/index.json", a.Auth(), MetrictankProxy("/metrics/index.json"))
 	m.Get("/graphite/metrics/index.json", a.Auth(), MetrictankProxy("/metrics/index.json"))
 	m.Any("/graphite/*", a.Auth(), GraphiteProxy)
+	m.Any("/prometheus/receive", a.Auth(), PrometheusWrite)
```
Comment (DanCech) [inline api/api.go]: not sure whether we should make this `/prometheus/write` since we're planning to add `/prometheus/read` later

Comment (woodsaj) [reply api/api.go]: I also prefer the term "write" here.

File: api/prometheus.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+				if l.Name == model.MetricNameLabel {
+					name = l.Value
+				} else {
+					tagSet = append(tagSet, l.Name+"="+l.Value)
+				}
+			}
+			if name != "" {
+				for _, sample := range ts.Samples {
+					md := metricPool.Get()
+					*md = schema.MetricData{
```
Comment (woodsaj) [inline api/prometheus.go]: You need to set the `Interval` field.

File: api/prometheus.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	"github.com/raintank/tsdb-gw/metricpool"
+	"github.com/raintank/worldping-api/pkg/log"
+	schema "gopkg.in/raintank/schema.v1"
+)
+
+var (
+	metricPool = metricpool.NewMetricDataPool()
+)
+
+func getSchemas(file string) (*conf.Schemas, error) {
```
Comment (woodsaj) [inline api/prometheus.go]: I dont see this getting called by anything.

**Comment by Dieterbe** [comment]
```
api/prometheus.go:7:2: cannot find package "github.com/gogo/protobuf/proto" in any of:
	/go/src/github.com/raintank/tsdb-gw/vendor/github.com/gogo/protobuf/proto (vendor tree)
	/usr/local/go/src/github.com/gogo/protobuf/proto (from $GOROOT)
	/go/src/github.com/gogo/protobuf/proto (from $GOPATH)
api/prometheus.go:9:2: cannot find package "github.com/grafana/metrictank/conf" in any of:
	/go/src/github.com/raintank/tsdb-gw/vendor/github.com/grafana/metrictank/conf (vendor tree)
	/usr/local/go/src/github.com/grafana/metrictank/conf (from $GOROOT)
	/go/src/github.com/grafana/metrictank/conf (from $GOPATH)
api/prometheus.go:10:2: cannot find package "github.com/prometheus/common/model" in any of:
	/go/src/github.com/raintank/tsdb-gw/vendor/github.com/prometheus/common/model (vendor tree)
	/usr/local/go/src/github.com/prometheus/common/model (from $GOROOT)
	/go/src/github.com/prometheus/common/model (from $GOPATH)
api/prometheus.go:11:2: cannot find package "github.com/prometheus/prometheus/prompb" in any of:
	/go/src/github.com/raintank/tsdb-gw/vendor/github.com/prometheus/prometheus/prompb (vendor tree)
	/usr/local/go/src/github.com/prometheus/prometheus/prompb (from $GOROOT)
	/go/src/github.com/prometheus/prometheus/prompb (from $GOPATH)
Exited with code 1
```

File: scripts/deploy_docker.sh [hunk 4]
```diff
@@ -8,4 +8,4 @@ cd ${DIR}
 VERSION=`git describe --always`
 
 docker push raintank/tsdb-gw:$VERSION
-docker push raintank/tsdb-gw:latest
+# docker push raintank/tsdb-gw:latest
```
Comment (Dieterbe) [inline scripts/deploy_docker.sh]: what does this have to do with prometheus?

File: api/prometheus.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+	var err error
+	flag.StringVar(&schemasConf, "prom-schemas-file", "/etc/storage-schemas.conf", "path to carbon storage-schemas.conf file for prom metrics")
+
+	schemas, err = getSchemas(schemasConf)
+	if err != nil {
+		log.Fatal(4, "failed to load schemas config. %s", err)
+	}
+}
+
+func getSchemas(file string) (*conf.Schemas, error) {
```
Comment (Dieterbe) [inline api/prometheus.go]: this function wrappes conf.ReadSchemas unnessarily. can just call conf.ReadSchemas directly instead of calling this function

File: api/prometheus.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+	schemas, err := conf.ReadSchemas(file)
+	if err != nil {
+		return nil, err
+	}
+	return &schemas, nil
+}
+
+func PrometheusWrite(ctx *Context) {
+	defer ctx.Req.Request.Body.Close()
+	if ctx.Req.Request.Body != nil {
```
Comment (Dieterbe) [inline api/prometheus.go]: put the defer after this check, otherwise can get nil panic.

File: api/prometheus.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+						Time:     (sample.Timestamp / 1000),
+						Mtype:    "gauge",
+						Tags:     tagSet,
+						OrgId:    ctx.OrgId,
+					}
+					md.SetId()
+					buf = append(buf, md)
+				}
+			} else {
+				log.Warn("prometheus metric received with empty name")
```
Comment (Dieterbe) [inline api/prometheus.go]: what does it mean when this happens? did the sender construct invalid data? if so, maybe a http 400 is in order?

Comment (jtlisi) [reply api/prometheus.go]: The name field of a prometheus metric can not be blank. I guess a 400 would make sense. However, does 1 poorly formatted metric require throwing out the entire payload?

Comment (Dieterbe) [reply api/prometheus.go]: it's IMHO the best we can do. better at least than dropping the data and not telling the client.

Comment (jtlisi) [reply api/prometheus.go]: agreed, maybe a counter is in order here

File: api/prometheus.go [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+)
+
+func PrometheusInit() {
+	if !*enabled {
+		return
+	}
+	log.Info("prometheus input enabled")
+	s, err := conf.ReadSchemas(*schemaFile)
+	if err != nil {
+		log.Fatal(4, "failed to load prometheus schemas config. %s", err)
```
Comment (Dieterbe) [inline api/prometheus.go]: either turn %s into %q or first do a check to confirm `*schemaFile != ""`
otherwise this would print only some whitespace and it'll be confusing

File: api/prometheus.go [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+	"github.com/raintank/tsdb-gw/metricpool"
+	"github.com/raintank/worldping-api/pkg/log"
+	schema "gopkg.in/raintank/schema.v1"
+)
+
+var (
+	metricPool = metricpool.NewMetricDataPool()
+	schemas    *conf.Schemas
+	schemaFile = flag.String("prom-schemas-file", "/etc/storage-schemas.conf", "path to carbon storage-schemas.conf file for prom metrics")
+	enabled    = flag.Bool("prometheus-enabled", false, "enable prometheus input")
```
Comment (Dieterbe) [inline api/prometheus.go]: consistency. either prom or prometheus for both

File: api/prometheus.go [hunk 10]
```diff
@@ -90,7 +93,9 @@ func PrometheusWrite(ctx *Context) {
 					buf = append(buf, md)
 				}
 			} else {
-				log.Warn("prometheus metric received with empty name")
+				log.Error(3, "prometheus metric received with empty name: %v", ts.String())
+				ctx.JSON(400, "metric written without name label")
```
Comment (Dieterbe) [inline api/prometheus.go]: not "written", that's confusing

File: api/prometheus.go [hunk 11]
```diff
@@ -16,17 +16,20 @@ import (
...
 
 var (
-	metricPool = metricpool.NewMetricDataPool()
-	schemas    *conf.Schemas
-	schemaFile = flag.String("prom-schemas-file", "/etc/storage-schemas.conf", "path to carbon storage-schemas.conf file for prom metrics")
-	enabled    = flag.Bool("prometheus-enabled", false, "enable prometheus input")
+	metricPool             = metricpool.NewMetricDataPool()
+	schemas                *conf.Schemas
+	schemaFile             = flag.String("prometheus-schemas-file", "/etc/storage-schemas.conf", "path to carbon storage-schemas.conf file for prom metrics")
+	prometheusWriteEnabled = flag.Bool("prometheus-enabled", false, "enable prometheus input")
```
Comment (Dieterbe) [inline api/prometheus.go]: is there a specific reason to make this configurable / default to false?
i personally don't really care but i don't see why we wouldn't just accept it by default. that's also what we discussed in the HM meeting on monday.

Comment (jtlisi) [reply api/prometheus.go]: I have no reason

Comment (Dieterbe) [reply api/prometheus.go]: ok so switch it to true by default then.

Comment (jtlisi) [reply api/prometheus.go]: One flag that I want to alter is the prometheus-schema-file. Since it technically is using the same file as graphite I would prefer them to use the same flag if that's ok

Comment (Dieterbe) [reply api/prometheus.go]: it's conceivable a client may have different schemas for carbon as for prometheus, so i would keep them separate.

**Comment by Dieterbe** [comment]
after publishing the data, you need to put the md's back into the metricPool.
(see carbon input)

File: api/prometheus.go [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+			log.Error(3, "Unmarshal Error, %v", err)
+			return
+		}
+
+		buf := make([]*schema.MetricData, 0)
+		for _, ts := range req.Timeseries {
+			var name string
+			var tagSet []string
+
+			_, s := schemas.Match(name, 0)
```
Comment (Dieterbe) [inline api/prometheus.go]: you're doing schemas.Match on name when it is still "" instead of after setting it to the proper value?

**Comment by Dieterbe** [comment]
I fixed the remaining bugs that I saw (please have a look at them). looks good to me now.
